pkg/reader/unserializer_cdx.go

```diff
@@ -13,11 +13,11 @@ import (
 	"github.com/bom-squad/protobom/pkg/sbom"
 )
 
-type UnserializerCDX14 struct{}
+type UnserializerCDX struct{}
```

Comment:
I started with this approach but ended up with just tons of copy-pasting.

We can compose different functionalities and rebuild the entire serializer/unserializer to be composed of pure utility functions, or use another layer for converting between CDX and protobom structures, but this ends up with having another layer of abstraction that still forces serializer/unserializer to convert on top of a common conversion object.

So if we embed `UnserializerCDX` into `UnserializerCDX15` 

And we want to use `common.ComponentToNode()` for all common parts, we must reimplement it again to support the new `ModelCard modelCard` from CDX1.5 to correctly parse into the sbom.Document.NodeList we created, while reiterating the entire SBOM component list.
This would not only make it less readable, but it will also make it less performant (think about large SBOMs)

Assuming compatibility with `cyclonedx-go`, my approach will serialize/unserialize all 1.x versions and support the conversion process for newer/deprecated features.

I agree we can build a better serializer/unserializer making them more "functional", but not with going deeper with object oriented.

Reply:
a more "generic" serializer build from more composable parts would make the most sense to you, but I still feel like its over-engineering for this specific stage we're at.

Consider something like:

```go
type Unserializer struct {
        bom cdx.BOM
        doc sbom.Document
	metadataFunc func(...)
}

type Option func(*Unserializer)

func (*Unserializer) defaultMetadata() {
	...
}

func WithLifecycle() Option {
	return func(u *Unserializer) {
             u.metadataFunc = func(...) {
		m := defaultMetadata(u.bom)
                 for _, lc := range *u.bom.Metadata.Lifecycles {
                    ...
                   m.doc.Lifecycles = ...
		}
	}
}

func NewUnserializer(opts ...Option) *Unserializer {
	u := &Unserializer{
		...
	}

	for _, opt := range opts {
		// Call the option giving the instantiated
		opt(u)
	}

	return u
}

func (u *Unserializer) Unserialize(...) {
      ....
      u.metadataFunc(...)
}

```

This was "easy" for Metadata, but breaking down `componentToNode` is 10x harder where we already check if a field exists. We already check if those fields are not nil 
```
if c.PackageURL != "" {
		node.Identifiers[int32(sbom.SoftwareIdentifierType_PURL)] = c.PackageURL
	}
``` 

We can do something more "generic" and create "generic" composable "path" based convertors, accepting "path" or type, but this means using `reflection` which is a no-go, in my opinion.

Reply:
I assume this would only be true for CDX v1.X and not for CDX v2 which would break compatibility.

Reply:
Also, this does not apply to SPDX, as they do not follow the same approach; they have dedicated structs/interfaces/functions for different spec versions. And I believe it would be best if we'd just *copy-paste* the serializer/unserializer, and reimplement it in an ad-hoc way.

you won't be able to embed a common serializer as each version uses different structs and imports.

Reply:
The main objective here is to have separate objects that let us swap the serializers and readers of any format for better implementations when we need it, all while reusing code and making sure the required data assignment decisions needed in each case can be done in isolation.

If this means having a type that does nothing but embed a single CDX parser, so be it. Same for writers. The requirement is simply that it makes a clear distinction on how it behaves with each format so that we don't read or assign fields from other versions when transcoding to the protobom documents.

>  we must reimplement it again to support the new ModelCard modelCard from CDX1.5

If we want to support this it could be fine, but someone needs to take on the work to ensure the model card is correctly disassembled into the graph it is representing. I would suggest not spending time on it just now as it is a ton of work and we have other more pressing issues.

Reply:
I understand; however, I would argue that since we use `cyclonedx-go` that doesn't differentiate between CDX versions, we would have to work hard on breaking the serializer/unserializer (for CDX1.x) into very small building blocks and rebuild the existing serializer without providing much value.

The compatibility layer is already baked into `cyclonedx-go`, if you decode a cdx1.4 SBOM you won't have a value for  Metadata.Lifecycle, it would be nil. 

I gave `ModelCard` as an example of a new prop that was recently added. if this was essential to SBOM and we want to support it in `protobom` we would have to convert all components to nodelist by using the "base/common" parser and then implement the new `ModelCard` support and convert them again, but now only for extending the proto with data from `ModelCard`

As you mention, we have more pressing issues, and in my opinion, It would be a waste of time trying to build a "common" parser/writer that is configurable, extendable, and performant. 

if you meant for the entire functionality to be implemented to the common cdx parser, and just make a wrapper that calls it, I would argue that this is just redundant and adds no value (because we don't add anything on top of the base parser), this is just an instance of a parser/writer configured, no need to create multiple files and types. 

```
type SerializerCDX14 struct {
	SerializerCDX
}

// Render is a wrapper on top of the general CDX serializer
func (s *SerializerCDX14) Render(_ options.Options, doc interface{}, wr io.Writer) error {
	// Call the global CycloneDX serializer method to render the doc
	return s.renderVersion(cdx.SpecVersion1_5, doc, wr)
}
```

Make a Serializer service, and instantiate it with options, specversions, encoding. 
The issue with embedding is that it makes both SerializerCDX and SerializerCDX14 incomplete, as neither implements the interface and is not very readable. 

I do see the point in instantiating different writers but when the renderVersion function is

```go
func (s *SerializerCDX) renderVersion(cdxVersion cdx.SpecVersion, doc interface{}, wr io.Writer) error {
	if doc == nil {
		return errors.New("document is nil")
	}

	encoder := cdx.NewBOMEncoder(wr, cdx.BOMFileFormatJSON)
	encoder.SetPretty(true)

	if err := encoder.EncodeVersion(doc.(*cdx.BOM), cdxVersion); err != nil {
		return fmt.Errorf("encoding sbom to stream: %w", err)
	}

	return nil
}
```

Creating 5 files with just:

```
func (s *SerializerCDX15) Render(_ options.Options, doc interface{}, wr io.Writer) error {
	// Call the global CycloneDX serializer method to render the doc
	return s.renderVersion(cdx.SpecVersion1_5, doc, wr)
}
...
func (s *SerializerCDX14) Render(_ options.Options, doc interface{}, wr io.Writer) error {
	// Call the global CycloneDX serializer method to render the doc
	return s.renderVersion(cdx.SpecVersion1_4, doc, wr)
}
...
func (s *SerializerCDX13) Render(_ options.Options, doc interface{}, wr io.Writer) error {
	// Call the global CycloneDX serializer method to render the doc
	return s.renderVersion(cdx.SpecVersion1_3 doc, wr)
}

```

Instead of a single:

```go 
serializer := NewSerializer(
   withCDX(specVersion)
   withEncoding(...)
)

Makes less sense.

Reply:
I just re-read my response, and I believe we can easily come to an agreement if I try making the serializer/writer part just more tight. 

Ill work on my proposal and we can evaluate the pros/cons.

Reply:
Also, this makes more sense for the `writer` but less for the reader, as the reader has only one single implementation and we can only configure the encoding of the target sbom.